### PR TITLE
Add unit tests for transformation from XML report to PMD and PMD/CPD report

### PR DESCRIPTION
**Pull request by jchen-parasoft**

### CODE REVIEW

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        reportDirPath = reportDirPath + "/static";
+        params.put("reports.location", "static/cpptest_professional_report.xml");
+        File testReport = new File(reportDirPath + "/cpptest_professional_report.xml");
+        File pmdReport = new File(reportDirPath + "/pmd-cpptest_professional_report.xml");
+        File pmdCpdReport = new File(reportDirPath + "/pmdCpd-cpptest_professional_report.xml");
+        String pmdReportRelativePath = reportDir.toURI().relativize(pmdReport.toURI()).getPath();
+        prepareDataForTestTransformation();
+
+        doNothing().when(buildProgressLogger).message(Mockito.isA(String.class));
+        doNothing().when(buildProgressLogger).logMessage(Mockito.isA(BuildMessage1.class));
```
Comment (jchen-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java]: This is to mock the function "_build.getBuildLogger().logMessage(DefaultMessagesInfo.createTextMessage(anyString))" in ParasoftFindingsBuildProcess.java. And the parameter value used by this function can be collected. But the value is a "BuildMessage1" entity. If to create a new entity to assert this, there is an error could be displayed, which is that assert value cannot be found. I have tried to mock "BuildMessage1" and return a value when creating a new entity. But it still  does not work. So I just assert the function has been called in the following assertation.

**Comment by yshi-parasoft** [comment]
Please put all properties files into a folder named "settings" all reports files into a folder named "reports".

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        Mockito.verify(buildProgressLogger, Mockito.atLeastOnce()).error(Mockito.isA(String.class));
+        Mockito.verify(buildProgressLogger, Mockito.atLeastOnce()).warning(Mockito.isA(String.class));
+        Mockito.verify(buildProgressLogger, Mockito.atLeastOnce()).buildFailureDescription(Mockito.isA(String.class));
+
+        Mockito.verify(buildProgressLogger).error("Unexpected report format: "+testReport.getAbsolutePath()+ ". \nPlease try recreating the report. If this does not resolve the issue, please contact Parasoft support.");
+        Mockito.verify(buildProgressLogger).warning("Skipping unrecognized report file: " + testReport.getAbsolutePath());
+        Mockito.verify(buildProgressLogger).buildFailureDescription("Failed to parse XML report");
+    }
+
+    private void prepareDataForTestTransformation() {
```
Comment (yshi-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java]: Change this method name to "setupMockedDataForTransformation".

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        doReturn(params).when(context).getRunnerParameters();
+        doReturn(buildProgressLogger).when(build).getBuildLogger();
+        doReturn(checkoutDir).when(build).getCheckoutDirectory();
+    }
+
+    private void clearGeneratedTestFile(File file) {
```
Comment (yshi-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java]: Change this method name to "deleteIfExists".

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    ParasoftFindingsBuildProcess parasoftFindingsBuildProcess;
+    @Mock
+    AgentRunningBuild build;
+    @Mock
+    BuildRunnerContext context;
+    @Mock
+    BuildProgressLogger buildProgressLogger;
+
+    Map<String, String> params = new HashMap<>();
+    String reportDirPath = "src/test/resources/reports";
```
Comment (yshi-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java:32]: Move the initialization for "reportDirPath" and "checkoutDir" to "setUp()"

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    @Test
+    public void test_transformSOATestReport() throws Throwable {
+        File junitReport = new File(reportDirPath + "/junit-SOAtest_report.xml");
+        File pmdReport = new File( reportDirPath + "/pmd-SOAtest_report.xml");
+
+        // Given
+        try {
+            File testReport = new File(reportDirPath + "/SOAtest_report.xml");
```
Comment (yshi-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java]: Please see if there's any room to refactor the process to read/process result and get build status. 
(For example, line 111 to line 120 here)

File: parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+        // When
+        BuildFinishedStatus status = parasoftFindingsBuildProcess.call();
+        int invalidReportCount = parasoftFindingsBuildProcess.getInvalidReportCount();
+
+        // Then
+        Assertions.assertFalse(pmdReport.exists());
+        Assertions.assertFalse(pmdCpdReport.exists());
+        Assertions.assertEquals(status.name(), "FINISHED_FAILED");
+        Assertions.assertEquals(invalidReportCount, 1);
```
Comment (yshi-parasoft) [inline parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java:198]: It's better that if we can prevent the following printout when running "mvn test" by configure the exception as expected in the test:
image

@rwang-parasoft What do you think

Comment (rwang-parasoft) [reply parasoft-findings-teamcity-agent/src/test/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcessTest.java:198]: Yes, I agree with you.

**Pull request merged**